File: cpp/nc-150/sliding-window/best_time_to_buy_and_sell_stock.cpp
#include <vector>

using namespace std;

/*
 * Author: Valerie Doan
 * Date: October 11, 2024
 * Problem: Best Time to Buy and Sell Stock
 * Level: Easy
 * Neetcode Course: Neetcode 150
 * Neetcode Topic: Sliding Window
 */

int maxProfit(vector<int> &prices)
{
    int profit = 0;
    int L = 0;

    for (int R = 0; R < prices.size(); R++)
    {
        while (prices[R] - prices[L] < 0)
        {
            L++;
        }

        profit = max(prices[R] - prices[L], profit);
    }

    return profit;
}

File: python/nc-150/sliding-window/best_time_to_buy_and_sell_stock.py
"""
Author: Valerie Doan
Date: November 11, 2024
Problem: Best Time to Buy and Sell Stock
Level: Easy
Neetcode Course: Neetcode 150
Neetcode Topic: Sliding Window
"""


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = L = 0

        for R in range(len(prices)):
            while prices[R] - prices[L] < 0:
                L += 1

            max_profit = max(prices[R] - prices[L], max_profit)

        return max_profit
